### PR TITLE
Allow rollback in evaluation status

### PR DESCRIPTION
**Pull request by Abriscioli**
@andresmoschini 

Hi Andrés! Here is my change to allow rolling back in the evaluation state. Could you review please?

### CODE REVIEW

File: source/CommonJobs/CommonJobs.Application.EvalForm/Commands/GetEvaluatorEmployeesCommand.cs [hunk 1]
```diff
@@ -50,7 +51,16 @@ public override List<EmployeeEvaluationDTO> ExecuteWithResult()
...
                     TemplateId = e.TemplateId,
-                    IsEditable = getEvaluationEditable(e)
+                    IsEditable = getEvaluationEditable(e),
+                    PosibleRevertActions = RevertEvaluationActionsHelper.GetPosibleRevertActions(
+                        e.Finished,
+                        e.OpenToDevolution,
+                        e.CompanyEvaluationDone,
+                        e.ResponsibleEvaluationDone,
+                        e.AutoEvaluationDone,
+                        e.EvaluatorEvaluationDone)
```
Comment (andresmoschini) [inline source/CommonJobs/CommonJobs.Application.EvalForm/Commands/GetEvaluatorEmployeesCommand.cs]: Since there are more than a _Evaluator Evaluation_ it is not clear what it means, what do you think about `AllEvaluatorsEvaluationDone` or `AnyEvaluatorsEvaluationDone`?

File: source/CommonJobs/CommonJobs.Application.EvalForm/Commands/GetEvaluatorEmployeesCommand.cs [hunk 2]
```diff
@@ -50,7 +51,16 @@ public override List<EmployeeEvaluationDTO> ExecuteWithResult()
...
+                        .Select(x => new PosibleRevertActions { ActionName = RevertEvaluationActionsHelper.MapRevertActionName(x), ActionValue = x.ToString() })
```
Comment (andresmoschini) [inline source/CommonJobs/CommonJobs.Application.EvalForm/Commands/GetEvaluatorEmployeesCommand.cs]: What do you think about to move this logic to `RevertEvaluationActionsHelper`?

File: source/CommonJobs/CommonJobs.Application.EvalForm/Commands/RevertEvaluationStateCommand.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
+﻿using CommonJobs.Application.EvalForm.Indexes;
+using CommonJobs.Domain.Evaluations;
+using CommonJobs.Infrastructure.RavenDb;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Web;
+using static CommonJobs.Application.EvalForm.Helper.RevertEvaluationActionsHelper;
```
Comment (andresmoschini) [inline source/CommonJobs/CommonJobs.Application.EvalForm/Commands/RevertEvaluationStateCommand.cs]: I think that `using static` is not compatible with previous VS versions, what do you think about to use a more conservative syntax?

File: source/CommonJobs/CommonJobs.Application.EvalForm/Commands/GetEvaluatorEmployeesCommand.cs (hunk 1 above)
Comment (Abriscioli) [inline source/CommonJobs/CommonJobs.Application.EvalForm/Commands/GetEvaluatorEmployeesCommand.cs]: `AnyEvaluatorsEvaluationDone` would be more accurate. I'll take that

File: source/CommonJobs/CommonJobs.Application.EvalForm/Commands/RevertEvaluationStateCommand.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+            var possibleActions = GetPosibleRevertActions(
+                evaluation.Finished,
+                evaluation.ReadyForDevolution,
+                isCompanyEvalFinished,
+                isResponsibleEvalFinished,
+                isAutoEvalFinished,
+                isEvaluatorEvalFinished);
+
+            if (!possibleActions.Contains(_action))
```
Comment (andresmoschini) [inline source/CommonJobs/CommonJobs.Application.EvalForm/Commands/RevertEvaluationStateCommand.cs]: What do you think about to move _this verification_ and `MapToAction()` logic inside the Helper?

Maybe the helper could be a non-static class, that receive  the parameters in the constructor (isFinished,  isReadyForDevolution, isCompanyEvalFinished, isResponsibleEvalFinished, isAutoEvalFinished, isEvaluatorEvalFinished) with two methods, one to get the complete list and another to validate if an action (string) is available.

Comment (andresmoschini) [inline source/CommonJobs/CommonJobs.Application.EvalForm/Commands/RevertEvaluationStateCommand.cs]: Oh, I see that you need the you need the _parsed action_, so maybe parsing logic could be outside the helper.

File: source/CommonJobs/CommonJobs.Application.EvalForm/Commands/RevertEvaluationStateCommand.cs (hunk 3 above)
Comment (Abriscioli) [inline source/CommonJobs/CommonJobs.Application.EvalForm/Commands/RevertEvaluationStateCommand.cs]: Yes, I agree

File: source/CommonJobs/CommonJobs.Application.EvalForm/Helper/RevertEvaluationActionsHelper.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            }
+
+            if (isReadyForDevolution)
+            {
+                yield return RevertAction.CancelDevolution;
+            }
+
+            if (!isFinished
+                && !isReadyForDevolution
+                // TODO: verify if it is possible to have a null on calificationsByType[CalificationType.Company]
```
Comment (andresmoschini) [inline source/CommonJobs/CommonJobs.Application.EvalForm/Helper/RevertEvaluationActionsHelper.cs]: does this TODO annotation still make sense?

File: source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsApiController.cs [hunk 6]
```diff
@@ -11,6 +11,7 @@
 using System.Linq;
 using System.Web;
 using System.Web.Mvc;
+using static CommonJobs.Application.EvalForm.Helper.RevertEvaluationActionsHelper;
```
Comment (andresmoschini) [inline source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsApiController.cs]: `using static`

File: source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsApiController.cs [hunk 7]
```diff
@@ -137,10 +139,44 @@ public JsonNetResult DeletePeriod(string period)
...
+        public JsonNetResult GetPosibleActions(string period, string userName)
+        {
+            var evaluationId = EmployeeEvaluation.GenerateEvaluationId(period, userName);
+            var evaluation = RavenSession.Load<EmployeeEvaluation>(evaluationId);
+            var califications = RavenSession.Advanced.LoadStartingWith<EvaluationCalification>(evaluationId + "/").ToList();
+            var isResponsibleEvalFinished = califications.Any(x => x.Owner == CalificationType.Responsible && x.Finished);
+            var isCompanyEvalFinished = califications.Any(x => x.Owner == CalificationType.Company && x.Finished);
+            var isAutoEvalFinished = califications.Any(x => x.Owner == CalificationType.Auto && x.Finished);
+            var isEvaluatorEvalFinished = califications.Any(x => x.Owner == CalificationType.Evaluator && x.Finished);
+            var actions = GetPosibleRevertActions(
```
Comment (andresmoschini) [inline source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsApiController.cs]: `GetPosibleRevertActions` is being used when you get the evaluations list, and also when you _revert an evaluation_, why is it need a third time?

File: source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsController.cs [hunk 8]
```diff
@@ -18,6 +18,7 @@
 using CommonJobs.Application.EvalForm.Indexes;
 using System.Web.Routing;
 using Raven.Client;
+using static CommonJobs.Application.EvalForm.Helper.RevertEvaluationActionsHelper;
```
Comment (andresmoschini) [inline source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsController.cs]: Unnecessary change

File: source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsController.cs [hunk 9]
```diff
@@ -173,8 +174,6 @@ public ActionResult Calification(string period, string username)
                 return new HttpStatusCodeResult(403, "Access Denied");
             }
 
-
-
```
Comment (andresmoschini) [inline source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsController.cs]: Unnecessary change

File: source/CommonJobs/CommonJobs.MVC.UI/Scripts/Evaluations/dashboard.js [hunk 10]
```diff
@@ -109,6 +109,39 @@
...
+            $('.content-evaluation').append($('.state-content-modal'));
+            $('.state-content-modal').hide();
+        };
+        this.revertEvaluationState = function (data) {
+            self.close;
+            viewmodel.isLoading(true);
+            var ajax = $.ajax("/Evaluations/api/RevertEvaluationState/" + self.evaluation.period + "/" + self.evaluation.userName + "?operation=" + data.ActionValue, {
+                type: "POST",
+                success: function (data) {
+                    viewmodel.isLoading(false);
```
Comment (andresmoschini) [inline source/CommonJobs/CommonJobs.MVC.UI/Scripts/Evaluations/dashboard.js]: Maybe you can keep the loading state, it will disappear when current page is unloaded.

File: source/CommonJobs/CommonJobs.MVC.UI/Scripts/Evaluations/dashboard.js [hunk 11]
```diff
@@ -330,4 +375,13 @@
             alert('Fallo interno. Por favor recargue la página.');
         });
     }
+
+    function getPosibleRevertActions(period, username) {
```
Comment (andresmoschini) [inline source/CommonJobs/CommonJobs.MVC.UI/Scripts/Evaluations/dashboard.js]: Is i being used?

**Comment by andresmoschini** [comment]
Nice work @Abriscioli, review done.

File: source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsApiController.cs (hunk 7 above)
Comment (Abriscioli) [inline source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsApiController.cs]: Rigth, we don't need it

File: source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsController.cs (hunk 8 above)
Comment (Abriscioli) [inline source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsController.cs]: Removed

File: source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsApiController.cs (hunk 6 above)
Comment (Abriscioli) [inline source/CommonJobs/CommonJobs.MVC.UI/Areas/Evaluations/Controllers/EvaluationsApiController.cs]: Removed. I'll use a conservative syntax

File: source/CommonJobs/CommonJobs.Application.EvalForm/Helper/RevertEvaluationActionsHelper.cs (hunk 5 above)
Comment (Abriscioli) [inline source/CommonJobs/CommonJobs.Application.EvalForm/Helper/RevertEvaluationActionsHelper.cs]: no, It doesn't. Removed

File: source/CommonJobs/CommonJobs.Application.EvalForm/Commands/RevertEvaluationStateCommand.cs (hunk 4 above)
Comment (Abriscioli) [inline source/CommonJobs/CommonJobs.Application.EvalForm/Commands/RevertEvaluationStateCommand.cs]: Moved validation, mapping and parsing to the RevertEvaluationActionHelper

File: source/CommonJobs/CommonJobs.MVC.UI/Scripts/Evaluations/dashboard.js (hunk 11 above)
Comment (Abriscioli) [inline source/CommonJobs/CommonJobs.MVC.UI/Scripts/Evaluations/dashboard.js]: No, it is no longer in use. Removing it

File: source/CommonJobs/CommonJobs.MVC.UI/Scripts/Evaluations/dashboard.js (hunk 10 above)
Comment (Abriscioli) [inline source/CommonJobs/CommonJobs.MVC.UI/Scripts/Evaluations/dashboard.js]: Right. Removing this line
